### PR TITLE
Add quadraturefunctions to visit datacollection [quadspace-visit-dev]

### PR DESCRIPTION
**Pull request by IdoAkkerman**
This PR adds  quadraturefunctions to the visit data collection.
At this stage the quadraturefunctions can be written to file and read again.
Visualisation capability in visit will be worked on in parallel.

See #1477.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1513](https://github.com/mfem/mfem/pull/1513) | @IdoAkkerman | @tzanio | @rcarson3 + @bslazarov | 05/31/20 | 07/02/20 | ⌛due 07/09/20 | |

### CODE REVIEW

**Comment by tzanio** [comment]
LGTM

File: fem/datacollection.cpp [hunk 1]
```diff
@@ -598,14 +605,28 @@ void VisItDataCollection::LoadFields()
       // TODO: 1) load parallel GridFunction on one processor
       if (serial)
       {
-         field_map.Register(it->first, new GridFunction(mesh, file), own_data);
+         if ((it->second).association == "nodes")
```
Comment (kennyweiss) [inline fem/datacollection.cpp:608]: What are the valid values for ``VisItFieldInfo.association``?
Looking at ``RegisterField`` and ``RegisterQField``, this looks right.

Perhaps some comments can be added here?
https://github.com/mfem/mfem/blob/d87f4e347c1332181263a8f90adb6e66db479d01/fem/datacollection.hpp#L388-L397

Comment (IdoAkkerman) [reply fem/datacollection.cpp:608]: In Visit there is already a hook for the `elements` instead of `qpoints` I introduced here.
So I think it might be better to adopt `elements`, which is currently not used, as far as I know.

Comment (IdoAkkerman) [reply fem/datacollection.cpp:608]: There is also the issue of the LOD. I would like to use that to tell Visit the integration order.
Or should I introduce a new parameter in the file to  pass this info to visit.
Then there is also the issue of mapping an integration routine with the appropriate refinement routine in visit. I am aiming for a minimal implementation with abundant checks for cases that will not work.

**Comment by IdoAkkerman** [comment]
I have added `GetRefinementLevelFromElems` to translate the number quad points to an LOD.
This gives a -1 if not match is found. In that case visit will display an element average instead of the individual quadrature points.

I added a `openGL` point distribution options, this allows the GL points to be beter visualised in Visit. Instead of a uniform subdivision of elements, the subdivision is governed by the distribution of quad points.

**Comment by rcarson3** [comment]
@IdoAkkerman could you provide us with a relatively simple example to test this on at least for the PR? Maybe something like where the quadrature data you're saving off is a projection of the mesh nodes on the quadrature points (a single direction of the project is fine as well so we can plot a scalar field in visit). You could probably use these lines as a starting point https://github.com/mfem/mfem/blob/master/tests/unit/fem/test_quadf_coef.cpp#L20-L72

File: fem/datacollection.cpp [hunk 2]
```diff
@@ -431,6 +428,27 @@ void VisItDataCollection::RegisterField(const std::string& name,
...
+void VisItDataCollection::RegisterQField(const std::string& name,
+                                         QuadratureFunction *qf)
+{
+   int LOD = -1;
+   Mesh *mesh = qf->GetSpace()->GetMesh();
+   for (int e=0; e<qf->GetSpace()->GetNE(); e++)
+   {
+      int locLOD = GlobGeometryRefiner.GetRefinementLevelFromElems(
+                      mesh->GetElementBaseGeometry(e),
+                      qf->GetElementIntRule(e).GetNPoints());
```
Comment (v-dobrev) [inline fem/datacollection.cpp:445]: Why not just use `IntegrationRule::GetOrder` to derive the LOD?

Comment (IdoAkkerman) [reply fem/datacollection.cpp:445]: There is unfortunately not a (very) clear connection between quadrature order and quadrature points. Moreover not each distribution of quadrature points is associated with a refinement, at least not how the refinements are programmed at this time.  Off course some of this "logic" could also be moved to visit, and LOD could simply be the order (or order +1) similar as for the FESpace,

Comment (IdoAkkerman) [reply fem/datacollection.cpp:445]: Moreover the quadrature order is ussually ~ twice as high as the order of the shapefunctions.

**Comment by IdoAkkerman** [comment]
@rcarson3 i have done that for debug purposes. But will add a case to the unit tests, this should also simplify debuging/testsuites on the VisIt side. I will work probably not have time to work on it  before the 22nd.

**Comment by IdoAkkerman** [comment]
@rcarson3 Was able to pulls something together in the last minute.

**Comment by IdoAkkerman** [comment]
Will de AFK till the 23th.

File: tests/unit/fem/test_quadf_coef.cpp [hunk 3]
```diff
@@ -156,6 +157,73 @@ TEST_CASE("Quadrature Function Coefficients",
       REQUIRE(output.Norml2() < tol);
    }
 
+   SECTION("Write and Read Scalar Quadraturefunction")
```
Comment (rcarson3) [inline tests/unit/fem/test_quadf_coef.cpp:160]: So the changes made to this unit test really don't belong in this file. They would be better suited in the /tests/unit/fem/test_datacollection.cpp file where the other data collection are. So, I would suggest reverting changes made here and moving them over to the other file.

File: tests/unit/CMakeLists.txt [hunk 4]
```diff
@@ -18,35 +18,6 @@ include_directories(BEFORE ${PROJECT_BINARY_DIR})
 include_directories(BEFORE ${CMAKE_CURRENT_SOURCE_DIR})
 
 set(UNIT_TESTS_SRCS
```
Comment (rcarson3) [inline tests/unit/CMakeLists.txt:20]: Why are all of these deleted? Please revert these changes back.

File: fem/intrules.hpp [hunk 5]
```diff
@@ -293,7 +294,8 @@ class Quadrature1D
       GaussLobatto    = 1,
       OpenUniform     = 2,  ///< aka open Newton-Cotes
       ClosedUniform   = 3,  ///< aka closed Newton-Cotes
-      OpenHalfUniform = 4   ///< aka "open half" Newton-Cotes
+      OpenHalfUniform = 4,  ///< aka "open half" Newton-Cotes
+      ClosedGL        = 5   // Closed Gauss Legendre
```
Comment (rcarson3) [inline fem/intrules.hpp:298]: Convert to a doxygen comment and make comment inline with the style used in the above comments

File: fem/fespace.hpp [hunk 6]
```diff
@@ -723,6 +723,12 @@ class QuadratureSpace
    /// Return the total number of quadrature points.
    int GetSize() const { return size; }
 
+   /// Returns the mesh
+   inline Mesh *GetMesh() const { return mesh; }
+
+   /// Returns number of elements in the mesh.
+   inline int GetNE() const { return mesh->GetNE(); }
+
```
Comment (rcarson3) [inline fem/fespace.hpp:731]: These should be nice improvements to life functions to ```QuadratureSpace```.

File: fem/datacollection.hpp [hunk 7]
```diff
@@ -445,6 +446,10 @@ class VisItDataCollection : public DataCollection
    /// Add a grid function to the collection and update the root file
    virtual void RegisterField(const std::string& field_name, GridFunction *gf);
 
+   /// Add a quadrature function to the collection and update the root file
+   virtual void RegisterQField(const std::string& q_field_name,
```
Comment (rcarson3) [inline fem/datacollection.hpp:450]: Another comment should be added that this currently is not usable within Visit to visualize quadrature data.

Based on the below unit test, if you do try to within Visit it fails with an error from MFEM saying: ```MFEM abort: input stream is not a FiniteElementSpace!```

File: tests/unit/CMakeLists.txt (hunk 4 above)
Comment (IdoAkkerman) [reply tests/unit/CMakeLists.txt:20]: Sorry that was sloppy of me... just deleted them to speed up testing.

File: tests/unit/fem/test_quadf_coef.cpp (hunk 3 above)
Comment (IdoAkkerman) [reply tests/unit/fem/test_quadf_coef.cpp:160]: Indeed don't know why I did not consider that. Will move, and change the test to better fit the datacollection unittest.

File: fem/intrules.hpp (hunk 5 above)
Comment (IdoAkkerman) [reply fem/intrules.hpp:298]: Will do so.

File: fem/fespace.hpp (hunk 6 above)
Comment (IdoAkkerman) [reply fem/fespace.hpp:731]: I do not understand the comment? I interpret this as positive feadback and not a request for change. Is that correct?
If so, thank you! If not could you eleborate?

File: fem/datacollection.hpp (hunk 7 above)
Comment (IdoAkkerman) [reply fem/datacollection.hpp:450]: Will make a statement that VisIt 3.12 does not work. A patch has been sent to VisIt developers.

File: fem/fespace.hpp (hunk 6 above)
Comment (rcarson3) [reply fem/fespace.hpp:731]: It was just positive feedback :) and nothing for you to fix.

**Comment by IdoAkkerman** [comment]
@rcarson3 thank you for your input for correctting the sloppyness and making improvements.  I made the changes as requested.

File: fem/datacollection.hpp [hunk 8]
```diff
@@ -445,6 +446,12 @@ class VisItDataCollection : public DataCollection
    /// Add a grid function to the collection and update the root file
    virtual void RegisterField(const std::string& field_name, GridFunction *gf);
 
+   /// Add a quadrature function to the collection and update the root file.
+   /** Visualisation of quadrature function is not supported in VisIt(3.12).
```
Comment (rcarson3) [inline fem/datacollection.hpp:450]: typo here should be Visualization

**Comment by tzanio** [comment]
Extending the deadline to this Friday...

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Merged in `next` for testing...
